bokeh/io.py

```diff
@@ -393,14 +394,21 @@ def _detect_filename(ext):
     None if the script could not be found (e.g. interactive mode).
     """
     import inspect
-    from os.path import isfile, dirname, basename, splitext, join
+    from os.path import isfile, dirname, basename, splitext, join, curdir
 
     frame = inspect.currentframe()
     while frame.f_back and frame.f_globals.get('name') != '__main__':
         frame = frame.f_back
 
     filename = frame.f_globals.get('__file__')
-    if filename and isfile(filename):
+
+    if filename is None:
+        return tempfile.NamedTemporaryFile().name
+
+    elif not os.access(dirname(filename) or curdir, os.W_OK | os.X_OK):
+        return tempfile.NamedTemporaryFile().name
+
+    elif isfile(filename):
         name, _ = splitext(basename(filename))
         return join(dirname(filename), name + "." + ext)
```

Comment:
I wonder if it is also worth checking whether the final filename exists and is not writable. Maybe the standard exception that python will raise if trying to write to an unwritable file is actually the clearest thing, though.